backend/tilavarauspalvelu/models/payment_order/queryset.py

```diff
@@ -16,15 +19,52 @@
 
 
 class PaymentOrderQuerySet(models.QuerySet):
-    def expired(self) -> Self:
-        older_than_minutes = settings.VERKKOKAUPPA_ORDER_EXPIRATION_MINUTES
-        expired_datetime = local_datetime() - datetime.timedelta(minutes=older_than_minutes)
+    def expired_direct_payments(self) -> Self:
+        now = local_datetime()
+        expiration = datetime.timedelta(minutes=settings.VERKKOKAUPPA_ORDER_EXPIRATION_MINUTES)
+        expires_at = now - expiration
 
         return self.filter(
             status=OrderStatus.DRAFT,
-            created_at__lte=expired_datetime,
+            created_at__lte=expires_at,
             remote_id__isnull=False,
         )
 
+    def expired_handled_payments(self) -> Self:
+        now = local_datetime()
+        expiration = datetime.timedelta(minutes=settings.VERKKOKAUPPA_ORDER_EXPIRATION_MINUTES)
+        expires_at = now - expiration
 
-class PaymentOrderManager(models.Manager.from_queryset(PaymentOrderQuerySet)): ...
+        return self.filter(
+            # Payment is overdue
+            models.Q(status=OrderStatus.PENDING)
+            & models.Q(handled_payment_due_by__lt=now)
+            # Payment never started, or payment expired
+            & (models.Q(remote_id__isnull=True) | models.Q(created_at__lte=expires_at))
+        )
+
+
+# Need to do this to get proper type hints in the manager methods, since
+# 'from_queryset' returns a subclass of Manager, but is not typed correctly...
+_BaseManager: type[models.Manager] = models.Manager.from_queryset(PaymentOrderQuerySet)  # type: ignore[assignment]
+
+
+class PaymentOrderManager(_BaseManager):
+    # Define to get type hints for queryset methods.
+    def all(self) -> PaymentOrderQuerySet:
+        return super().all()  # type: ignore[return-value]
+
+    def cancel_expired_payments_in_verkkokauppa(self) -> None:
+        for payment_order in self.all().expired_direct_payments():
+            # Do not update PaymentOrder status if an error occurs
+            with suppress(GetPaymentError, CancelOrderError), transaction.atomic():
+                payment_order.actions.refresh_order_status_from_webshop()
+                if payment_order.status == OrderStatus.EXPIRED:
+                    payment_order.actions.cancel_together_with_verkkokauppa()
+
+        for payment_order in self.all().expired_handled_payments():
+            # Do not update PaymentOrder status if an error occurs
+            with suppress(GetPaymentError, CancelOrderError), transaction.atomic():
+                payment_order.actions.refresh_order_status_from_webshop()
+                if payment_order.status == OrderStatus.EXPIRED:
+                    payment_order.actions.cancel_together_with_verkkokauppa()
```

Comment:
The _payment order_ cancellation can, I think, be omitted (both here and above) since the expiration time set into the webshop order should now prevent them from proceeding further in the webshop payment flow.
Then `OrderStatus.EXPIRED` will also remain as the final status (unless user explicitly cancels or denies the reservation after the expiration but before the `handle_unfinished_reservations` task is run), which is proper.

Action could then be named e.g. `update_expired_payments_in_verkkokauppa` -- or come to think of it, even more generally as `update_payment_statuses` and the task renamed correspondingly, since (a), unpaid handled may not exist in webshop and (b) it secondarily also ensures that the status of an order is eventually reconciled, in case webhook fails.